### PR TITLE
os/bluestore: replicate bdev label 

### PR DESCRIPTION
**Pull request by pereman2**
Currently this is wip as I'm working on trying to replicate bluestore's label while not messing existing allocations on where replicated label will be placed in case of backports.

The idea is to replicate bluestore's label on different offsets like: 0, 1GiB 10GiB and 1TiB. Currently, only /block is being replicated but it is expected to expand this onto bluefs if possible.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -9168,6 +9243,15 @@ int BlueStore::_fsck(BlueStore::FSCKDepth depth, bool repair)
   if (r < 0) {
     return r;
   }
+
+  if (repair) {
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:9247]: One cannot repair bdev label in fsck.
To run fsck both bluefs and bluestore must be properly initialized.
Should bdev label be broken initialization will fail.

Repair of bdev label must be an external action.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -5415,62 +5417,126 @@ int BlueStore::_write_bdev_label(CephContext *cct,
...
-    t.substr_of(bl, 0, p.get_off());
-    crc = t.crc32c(-1);
-    decode(expected_crc, p);
-  }
-  catch (ceph::buffer::error& e) {
-    derr << __func__ << " unable to decode label at offset " << p.get_off()
-	 << ": " << e.what()
-	 << dendl;
-    return -ENOENT;
+int BlueStore::_read_bdev_label(CephContext* cct, const string &path,
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:5494]: _read_bdev_label works now also as a recovery, let's discuss it on BSUE.

Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:5550]: as per the meeting, we decided to only fix on fsck

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -5435,9 +5447,21 @@ void BlueStore::_close_path()
...
+      existing_allocation_in_offset = false;
+      }
+      });
+
+  if (!existing_allocation_in_offset) {
+    alloc->init_rm_free(offset, min_alloc_size);
+  }
+}
+
+int BlueStore::_replicate_bdev_label(const string &path, bluestore_bdev_label_t label, bool alloc_available) {
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:5464]: Why `_replicate_bdev_label` is separate from `_write_bdev_label`?
All of them should always contain the same data.
I think that the additional locations of label should be handled inside `_write_bdev_label()`.

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -5456,63 +5480,122 @@ int BlueStore::_write_bdev_label(CephContext *cct,
...
-    return -ENOENT;
+int BlueStore::_read_bdev_label(CephContext* cct, const string &path,
+				bluestore_bdev_label_t *label, bool replicated)
+{
+  dout(10) << __func__ << dendl;
+  bufferlist bl;
+  string error;
+  bluestore_bdev_label_t tmp_label;
+  size_t crcs_failed = 0;
+  for (uint64_t offset : bdev_label_offsets) {
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:5558]: This function does not check if all labels are exactly the same.
Having multiple different labels is also an error.

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -5456,63 +5480,122 @@ int BlueStore::_write_bdev_label(CephContext *cct,
...
-    decode(expected_crc, p);
-  }
-  catch (ceph::buffer::error& e) {
-    derr << __func__ << " unable to decode label " << path.c_str()
-         << " at offset " << p.get_off()
-	 << ": " << e.what()
-	 << dendl;
-    return -ENOENT;
+int BlueStore::_read_bdev_label(CephContext* cct, const string &path,
+				bluestore_bdev_label_t *label, bool replicated)
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:5551]: The name 'replicated' is confusing for me.
"false" = require all copies are ok
"true" = be satisfied if one copy is ok
How about "recovery_mode"?

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -7252,8 +7335,16 @@ int BlueStore::mkfs()
...
 
   reserved = _get_ondisk_reserved();
+
   alloc->init_add_free(reserved,
     p2align(bdev->get_size(), min_alloc_size) - reserved);
+
+  {
+    string p = path + "/block";
+    bluestore_bdev_label_t label;
+    _read_bdev_label(cct, p, &label);
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:7345]: Cannot read & replicate here; there are subsequent calls to `write_meta()`

File: src/os/bluestore/BlueStore.cc [hunk 7]
```diff
@@ -9195,12 +9292,28 @@ int BlueStore::_fsck(BlueStore::FSCKDepth depth, bool repair)
       depth == FSCK_SHALLOW ? " (shallow)" : " (regular)")
     << dendl;
 
+  if (repair) {
+    string p = path + "/block";
+    bluestore_bdev_label_t label;
+    _read_bdev_label(cct, p, &label, true);
+    _replicate_bdev_label(p, label, false);
+  }
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:9300]: This cannot succeed if 0 labels were correct.

File: src/os/bluestore/BlueStore.cc [hunk 8]
```diff
@@ -9195,12 +9292,28 @@ int BlueStore::_fsck(BlueStore::FSCKDepth depth, bool repair)
...
+
+  for (uint64_t offset : bdev_label_offsets) {
+    if (offset >= bdev->get_size()) {
+      break;
+    }
+    _unalloc_bdev_label_offset(offset);
+  }
+
+
   auto close_db = make_scope_guard([&] {
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:9317]: At location: https://github.com/ceph/ceph/pull/53095/files#diff-6f3b1a5dcf5313dc06f71ba4997ee28ec219ecb5633a728b916465c684f1d87dL9308
is the place where the bdev replicas should be marked.

File: src/os/bluestore/BlueStore.cc [hunk 9]
```diff
@@ -5435,9 +5447,21 @@ void BlueStore::_close_path()
   path_fd = -1;
 }
 
-int BlueStore::_write_bdev_label(CephContext *cct,
-				 const string &path, bluestore_bdev_label_t label)
-{
+void BlueStore::_unalloc_bdev_label_offset(uint64_t offset) {
+  ceph_assert(alloc);
+  bool existing_allocation_in_offset = true;
+  alloc->foreach([&](uint64_t allocation_offset, uint64_t allocation_length) {
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:5453]: Naming is just bad.
allocation_offset = means free region starts here
allocation_length = free region has this size

File: src/os/bluestore/BlueStore.cc [hunk 10]
```diff
@@ -142,6 +142,18 @@ const string PREFIX_ZONED_CL_INFO = "G";  // (per-zone cleaner metadata)
...
+// were already used so labels won't exist there.
+const vector<uint64_t> bdev_label_offsets = {0, 1073741824, 107374182400, 1099511627776};
+
+// write a label in the first block.  always use this size.  note that
+// bluefs makes a matching assumption about the location of its
+// superblock (always the second block of the device).
+#define BDEV_LABEL_BLOCK_SIZE  4096
+
+// reserve: label (4k) + bluefs super (4k), which means we start at 8k.
+#define SUPER_RESERVED  8192
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:155]: The change DB_SUPER_RESERVED -> SUPER_RESERVED is very good.
It has nothing to do with DB.

**Comment by pereman2** [comment]
jenkins test make check

File: src/os/bluestore/BlueStore.cc [hunk 11]
```diff
@@ -142,6 +142,10 @@ const string PREFIX_ZONED_CL_INFO = "G";  // (per-zone cleaner metadata)
 
 const string BLUESTORE_GLOBAL_STATFS_KEY = "bluestore_statfs";
 
+// Label offsets where they might be replicated. It is possible on previous versions where these offsets
+// were already used so labels won't exist there.
+const vector<uint64_t> bdev_label_offsets = {0, 1073741824, 107374182400, 1099511627776};
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:147]: what's about making this configurable through regular ceph settings?

Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:147]: @aclamk What do you think about this?

Comment (aclamk) [reply src/os/bluestore/BlueStore.cc:147]: Caph settings can change, but to locate bdev label copies it is best to have them in fixed, predictable location.
I do not think that having this configurable brings additional value.

**Comment by aclamk** [comment]
I feel like there will be a problem when we upgrade from previous version.
We should have something in bdev label that would tell us that writing multiple label copies is expected.
So, on upgraded OSDs it will work as before and only new deployments would have the multi copies.

**Comment by pereman2** [comment]
jenkins test make check

**Comment by ifed01** [comment]
@pereman2 - could you please provide some design overview behind your implementation so we can discuss the approach.

I've been thinking about something like the following but as far as I can see your approach is rather different:
1. Introduce epoch field to bdev label starting at 1. Epoch is just a counter incremented on each writing to bdev. Legacy labels to have zero epoch which is constant.
2. on mkfs - labels are written to multiple predefined locations with epoch = 1
3. on (first?) label reading - we read the label0 at offset 0, if epoch is zero - fall back to legacy single label mode. If label has got a non-zero epoch or label0 is corrupted (undecodable/bad crc/wrong cluster id/disk read error/etc) - read labels from all the other potential locations and report/note corrupted ones as per above. Among the valid labels the latest one is selected as active given epoch field and all further reads occur against it - alternatively we can repeat the above procedure each time. If no valid labels have been found - OSD fails.
4. on mount - if multilabel mode is discovered - unconditionally tag all label locations as non-free in allocator. 
5. on (or possibly after) umount - if multilabel mode has been in use - mark label locations as free as we do for BlueFS and persist allocator map in this state.
6. on label write - if new multilabel mode has been determined - write multiple labels using the active label instance (+updates) values and incremented epoch. Disk write errors to be reported to OSD log (and may be even as OSD alert) but shouldn't impact OSD functioning unless all the labels are unwritable. In legacy mode - just a single label0 to be written with zero epoch.

Alternatively (and apparently more in line with your approach) we can fail on step 3 if some label is invalid and request fsck to fix that but there is a drawback in this approach - some label locations could be permanently damaged due to e.g. disk errors - this will effectively forbid OSD from starting as the following loop (read - error - fsck - read - error - ...) occurs.
@aclamk - your thoughts are welcome too ;)

File: src/os/bluestore/BlueStore.cc [hunk 12]
```diff
@@ -5435,9 +5439,21 @@ void BlueStore::_close_path()
   path_fd = -1;
 }
 
-int BlueStore::_write_bdev_label(CephContext *cct,
-				 const string &path, bluestore_bdev_label_t label)
-{
+void BlueStore::_unalloc_bdev_label_offset(uint64_t offset) {
+  ceph_assert(alloc);
+  bool existing_allocation_in_offset = true;
+  alloc->foreach([&](uint64_t extent_offset, uint64_t extent_length) {
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:5445]: And most likely we can avoid this at all if we rely on bdev content to determine new multilabel mode and unconditionally mark relevant locations as non-free on mount.

**Comment by pereman2** [comment]
> @pereman2 - could you please provide some design overview behind your implementation so we can discuss the approach.
> 
> I've been thinking about something like the following but as far as I can see your approach is rather different:
> 
> 1. Introduce epoch field to bdev label starting at 1. Epoch is just a counter incremented on each writing to bdev. Legacy labels to have zero epoch which is constant.
> 2. on mkfs - labels are written to multiple predefined locations with epoch = 1
> 3. on (first?) label reading - we read the label0 at offset 0, if epoch is zero - fall back to legacy single label mode. If label has got a non-zero epoch or label0 is corrupted (undecodable/bad crc/wrong cluster id/disk read error/etc) - read labels from all the other potential locations and note valid/corrupted (as per above) ones. Among the valid labels the latest one is selected as active given epoch field and all further reads occur against it. If multilabel mode has been discovered - all the corrupted labels to be reported to log and possibly raise an OSD alert. If no(!) valid labels have been found - OSD fails to start.
> 4. Valid active label to be kept in RAM rather than re-read from disk each time which is different from what we have had so far IIRC, Alternatively we can repeat the above procedure on each read request but IMO this could spoil OSD log in case of any troubles.
> 5. on mount - if multilabel mode is discovered - unconditionally tag all label locations as non-free in allocator.
> 6. on (or possibly after) umount - if multilabel mode has been in use - mark label locations as free as we do for BlueFS and persist allocator map in this state.
> 7. on label write - if new multilabel mode has been determined - write multiple labels using the active label instance (+updates) values and incremented epoch. Disk write errors to be reported to OSD log (and may be even as OSD alert) but shouldn't impact OSD functioning unless all the labels are unwritable. In legacy mode - just a single label0 to be written with zero epoch.
> 
> Alternatively (and apparently more in line with your approach) we can fail on step 3 if some label is invalid and request fsck to fix that but there is a drawback in this approach - some label locations could be permanently damaged due to e.g. disk errors - this will effectively forbid OSD from starting as the following loop (read - error - fsck - read - error - ...) occurs. @aclamk - your thoughts are welcome too ;)

In general what you are proposing looks more or less like our first of many ideas. The key idea behind committed changes is to only deal with multiple labels on upgraded OSDs. If you upgraded and mount again `replicated_label`, without rerunning `mkfs`, boolean is checked to see if it was mounted with multiple labels -- we could change that to epoch just fine. 


Right now when we mount we read first label (`replicated_label == 1 || epoch == 1`) if it's corrupted we just don't mount and communicate it to later perform a fsck repair that will take care of fixing corrupted labels. In summary, we decided not to mount BlueStore if the first label is corrupted.

File: src/os/bluestore/BlueStore.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:5445]: Agree, I missed a procedure like that. How hard would it be to implement on all allocators? Are the some that we could skip?

File: src/os/bluestore/BlueStore.cc [hunk 13]
```diff
@@ -5456,65 +5472,161 @@ int BlueStore::_write_bdev_label(CephContext *cct,
...
-    derr << __func__ << " failed to write to " << path
-	 << ": " << cpp_strerror(r) << dendl;
-    goto out;
+  int r = 0;
+  for (uint64_t offset : bdev_label_offsets) {
+    if (alloc_available) {
+      if (offset >= bdev->get_size()) {
+        break;
+      }
+      _unalloc_bdev_label_offset(offset);
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/os/bluestore/BlueStore.cc:5481]: The idea is to only replicate only in OSDs  that just ran mkfs. We only treat these offsets as available on `replicated_label` ObjectStores:

```c++
    int r = read_meta("replicated_label", &value);
    if (r == 0 && value == "yes") {
      int r = _replicate_bdev_label(p, label, false);
```

File: src/os/bluestore/BlueStore.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:5445]: This is doable but for sure requires some efforts to maintain this functionality in every allocator (and cover with proper testing). Let's postpone this though till we agree on the overall design. May be we'll don't need it at all

**Review by ifed01 — CHANGES_REQUESTED** [review]
I'm still skeptical about the overall design - primarily I don't like extent at offset zero being a key stone - persistent disk error at it will make OSD unrecoverable - which IMO is a design flaw. And that's the issue which I really observed a few times in the field. 

Persistent disk errors at other label locations aren't well handled IMO for now as well. Could be fixed by proper implementation though.

Apart from that there are a few label corruption/recovery scenarios which aren't properly implemented as well - undecodable label or label mismatch fields are the examples. Definitely more thorough unit testing is required.

And finally I'd really like to have more options to control the feature:
1. Enable/disable label replication on mkfs. This will help in both working around some potential issues in the field and have better testing coverage - e.g. we can't implement "upgrade" test cases at store_test right now.
2. Enable/disable label replication in _write_bdev_label/_replicate_bdev_label even when it's been originally enabled on mkfs. Might be helpful to work around potential implementation issues in the field.

File: src/os/bluestore/BlueStore.cc (hunk 3 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:5456]: IMO  better then to have _write_bdev_label  being responsible specifically for writing labels to multiple destinations and to introduce another function to tag label's extents unallocated?

File: src/os/bluestore/BlueStore.cc [hunk 14]
```diff
@@ -9195,12 +9332,44 @@ int BlueStore::_fsck(BlueStore::FSCKDepth depth, bool repair)
...
+        derr << __func__ << " fsck error: all block device labels corrupted" << dendl;
+      }
+      return r;
+    }
+  }
+  if (repair) {
+    string value;
+    int r = read_meta("replicated_label", &value);
+    if (r == 0 && value == "yes") {
+      int r = _replicate_bdev_label(p, label, false);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:9358]: Please clarify what label do we use here as a source for the recovery? Presumably the one which we managed to read successfully above. But what if there exist two different but both valid labels, e.g. after an incomplete previous replication? Which one to choose? IMO that's another design flaw. We really need epoch/timestamp/seqno in the label to distinguish the latest replica.

**Pull request closed without merging**